Issue #26518: Graceful errors function_range handling for unsolvable critical point sets.

khalidbagus opened the pull request:
### Graceful errors function_range handling for unsolvable critical point sets.

This PR addresses issues in determining the range of functions when the derivative cannot be solved symbolically. In particular, it:

- Improves `function_range` so that if the critical-point equation (f′(x) = 0) remains unsolvable (i.e. yields a non-iterable ConditionSet or similar), it raises a detailed `NotImplementedError`. The error message now includes the derivative expression and suggests numerical methods or monotonicity analysis as alternatives.

This change fixes the issues observed with functions like `Abs((log(x))**(1/3))` and the more complex `log(1+x)/(x**(1/3)*(2+x))`.  
  
Fixes: #26518  
See also: [GitHub issue discussion](https://github.com/sympy/sympy/issues/26518)


#### Brief description of what is fixed or changed
- **function_range:**  
  Improved error handling so that if critical points remain unsolvable (even after simplification), a detailed `NotImplementedError` is raised. The error message now includes the unsolvable derivative expression and recommends alternative approaches.

#### Other comments
These changes help prevent low-level TypeErrors (such as those arising from attempting to iterate a ConditionSet) and give users clear guidance when analytic methods fail.

<!-- BEGIN RELEASE NOTES -->
* calculus
  * Improved `function_range` to detect unsolvable critical point sets and raise a detailed error message with suggestions for numerical or monotonicity analysis.
<!-- END RELEASE NOTES -->

Commit:
Issue #26518: Enhance solveset and function_range handling

- Added optional parameter 'simplify_derivative' to solveset with proper logic to retry derivative simplification when the initial solving returns unsolvable results.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError, suggesting numerical or monotonicity analysis.

Commit:
Fix test failures: update handling of critical points in function_range

khalidbagus committed:
Updated mailmap

khalidbagus committed:
Updated mailmap

oscarbenjamin commented on sympy/solvers/solveset.py:
```diff
@@ -2334,7 +2334,7 @@ def add_type(lhs, rhs, symbol, domain):
 
 
-def solveset(f, symbol=None, domain=S.Complexes):
+def solveset(f, symbol=None, domain=S.Complexes, simplify_derivative=False):
```
We don't want to add a parameter to `solveset` for this. The caller can just call `simplify` if they want.

khalidbagus replied:
Since `function_range` has internal logic to use `solveset` to find critical points for given function — then the caller couldn't call `simplify` from outside `function_range` to impose simplification just for finding critical points of given function derivative.

Option if `simplify_derivative` shouldn't be added ->
Then i suggest `function_range`  should just resort to exit gracefully with message that explicit critical points couldn't be found because the derivative of given function is unsimplified in which could involving radicals or transcendental functions.

What do you think?

khalidbagus replied:
Like this (for example equation shown in Issue #26518 Abs(((log(x))**Rational(1, 3))) )
```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
Cell In[3], line 7
      4 x = Symbol('x', real=True) 
      5 func_expr = Abs(((log(x))**Rational(1, 3)))
----> 7 function_range(func_expr.simplify(), x, Interval.Lopen(0, 1))

File ~\OneDrive\Documents\opensource\sympy\sympy\calculus\util.py:268, in function_range(f, symbol, domain)
    261 # Check if the result is unsolvable or non-iterable
    262 if (isinstance(critical_points, ConditionSet) or
    263     (isinstance(critical_points, Complement) and 
    264     any(isinstance(sub, ConditionSet) for sub in critical_points.args)) or
    265     (isinstance(critical_points, Union) and 
    266     any(isinstance(sub, ConditionSet) or not iterable(sub) for sub in critical_points.args)) or
    267     not iterable(critical_points)):
--> 268     raise NotImplementedError(
    269         f"Unable to find explicit critical points for the derivative of given function:\n\n  {f.diff(symbol)} = 0\n\n"
    270         "This may be due to the unsimplified form of the derivative (e.g. involving radicals or transcendental functions)."
    271     )
    272 if isinstance(critical_points, ImageSet):
    273     raise NotImplementedError(
    274         f"Found an infinite number of critical points for the derivative:\n\n  {f.diff(symbol)} = 0\n\n"
    275         "This indicates a periodic function; consider restricting the domain."
    276     )

NotImplementedError: Unable to find explicit critical points for the derivative of given function:

  ((log(x/sign(x))*sin(atan2(arg(x), log(x/sign(x)))/3)*sign(x)**2/(3*x*(log(x/sign(x))**2 + arg(x)**2)**(5/6)) - cos(atan2(arg(x), log(x/sign(x)))/3)*arg(x)*sign(x)**2/(3*x*(log(x/sign(x))**2 + arg(x)**2)**(5/6)))*(log(x/sign(x))**2 + arg(x)**2)**(1/6)*sin(atan2(arg(x), log(x/sign(x)))/3) + (log(x/sign(x))*cos(atan2(arg(x), log(x/sign(x)))/3)*sign(x)**2/(3*x*(log(x/sign(x))**2 + arg(x)**2)**(5/6)) + sin(atan2(arg(x), log(x/sign(x)))/3)*arg(x)*sign(x)**2/(3*x*(log(x/sign(x))**2 + arg(x)**2)**(5/6)))*(log(x/sign(x))**2 + arg(x)**2)**(1/6)*cos(atan2(arg(x), log(x/sign(x)))/3))*sign(log(x)**(1/3))/log(x)**(1/3) = 0

This may be due to the unsimplified form of the derivative (e.g. involving radicals or transcendental functions).
```

khalidbagus committed:
Enhance function_range for unsolvable derivatives

- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See https://github.com/sympy/sympy/issues/26518 for discussion.

khalidbagus committed:
Updated mailmap

khalidbagus committed:
Enhance function_range for unsolvable derivatives

- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See https://github.com/sympy/sympy/issues/26518 for discussion.

khalidbagus committed:
Enhance function_range for unsolvable derivatives

- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See https://github.com/sympy/sympy/issues/26518 for discussion.

khalidbagus committed:
Updated mailmap

khalidbagus committed:
Updated mailmap

khalidbagus committed:
author: add khalidbagus to .mailmap

khalidbagus committed:
Graceful errors function_range handling

- Added timeout handler for the critical point set solver in function_range.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See https://github.com/sympy/sympy/issues/26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets.

- Added timeout handler for the critical point set solver in function_range.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See https://github.com/sympy/sympy/issues/26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added timeout handler for the critical point set solver in function_range.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes sympy#26518
See sympy#26518 for discussion.

khalidbagus committed:
test

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added timeout handler for the critical point set solver in function_range.
- Introducting fprime = f.diff(symbol) that is stored for the purpose of the usage of new _SolvetTimeoutError method
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See #26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added timeout handler for the critical point set solver in function_range.
- Introducting fprime = f.diff(symbol) that is stored for the purpose of the usage of new _SolvetTimeoutError method.
- Seconds timeout is set for 10 seconds to compute solveset of given derivative in function_range routine.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See #26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added an iteration limit safeguard in function_range to abort solveset calls that exceed 1000 iterations.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See #26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added an iteration limit safeguard in function_range to abort solveset calls that exceed 1000 iterations.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See #26518 for discussion.

khalidbagus committed:
Graceful errors function_range handling for unsolvable critical point sets

- Added an iteration limit safeguard in function_range to abort solveset calls that exceed 1000 iterations.
- Improved function_range to catch unsolvable critical point sets and raise a detailed NotImplementedError.
Fixes #26518
See #26518 for discussion.
